src/bn256/engine.rs

```diff
@@ -15,15 +15,18 @@ use pairing::{Engine, MillerLoopResult, MultiMillerLoop, PairingCurveAffine};
 use rand_core::RngCore;
 use subtle::{Choice, ConditionallySelectable, ConstantTimeEq};
 
+/// The seed U of generating the curve BN256
 pub const BN_X: u64 = 4965661367192848881;
 
-// 6U+2 for in NAF form
+/// 6U+2 for in NAF form
```

Comment:
I also don't know exact reason. 
But, I found that there was similar discussion about NAF representation.
https://github.com/ethereum/go-ethereum/issues/27957

In addition, I found that there are several sources for `6u + 2 NAF` value.
https://github.com/cloudflare/bn256/blob/master/optate.go#L115
https://github.com/AntelopeIO/bn256/blob/4370b6964a4fccc599c42fa7449fb042e59372c5/src/optate.h#L142
https://github.com/randombit/pairings.py/blob/master/bn256.py#L121
https://github.com/matter-labs/pairing/blob/master/src/bn256/mod.rs#L28
Seems that they use different NAF algorithm for representation.

Reply:
I have found some answers.

**Simple answer:**
This `SIX_U_PLUS_2_NAF` is not real `NAF` representation.
At least, it includes the consecutive non-zero entries.
The real `SIX_U_PLUS_2_NAF` presentation is:
```
pub const SIX_U_PLUS_2_NAF: [i8; 66] = [
    0, 0, 0, 1, 0, 1, 0, -1, 0, 0, -1, 0, 0, 0, 1, 0, 0, -1, 0, -1, 0, 0, 0, 1, 0, -1, 0, 0, 0, 0,
    -1, 0, 0, 1, 0, -1, 0, 0, 1, 0, 0, 0, 0, 0, -1, 0, 0, -1, 0, 1, 0, -1, 0, 0, 0, -1, 0, -1, 0,
    0, 0, 1, 0, -1, 0, 1,
];
```
Even though the current `SIX_U_PLUS_2_NAF` is correct for computation, it is a borrow from other repos([go-ethereum](https://github.com/ethereum/go-ethereum/blob/5976e58415a633c24a0d903e8a60a3780abdfe59/crypto/bn256/cloudflare/optate.go#L115-L118))
I still don't know what algo is used for this "NAF" representation.

**Long answer:**
First of all, the `bn256` curve in `halo2curves` repo is, in essence, `bn254`(or `alt_bn_128`) curve, where `u = 4965661367192848881`.
The devs found & discussed the issue [here](https://github.com/ethereum/go-ethereum/issues/21595).
Also, they resolved the issue in [this PR](https://github.com/ethereum/go-ethereum/pull/21836).
If you want more information about `bn254` curve, you should read [this](https://hackmd.io/@jpw/bn254).

Many projects([go-ethereum](https://github.com/ethereum/go-ethereum/blob/5976e58415a633c24a0d903e8a60a3780abdfe59/crypto/bn256/cloudflare/optate.go#L115-L118), [matter-labs](https://github.com/matter-labs/pairing/blob/d24f2c5871089c4cd4f54c0ca266bb9fef6115eb/src/bn256/mod.rs#L29-L38), [Antelope](https://github.com/AntelopeIO/bn256/blob/main/src/optate.h#L142-L145)) use the `bn256` curve, same as `halo2curves`.
They all use the same `SIX_U_PLUS_2_NAF` value(weird NAF from same `u` value).
Hence, I believe that the current `SIX_U_PLUS_2_NAF` could've been "borrowed" from other project.
(@kilic can you give some info about this?)

The real NAF representation of `SIX_U_PLUS_2_NAF` can be computed with cloudflare tool. 
We can find more details & code [here](https://github.com/cloudflare/bn256/issues/39). (You need to use the correct `u` value for computation)


After all of this inquiry, I think we have 2 options:
 - keep the current `SIX_PLUS_2_NAF`
 - use the real NAF representation 

Since they are all valid for computation, I think we can determine which one to use.  
In case of 1st option, it is good for compatibility.
In case of 2nd option, it does not bring confusion we had.